### PR TITLE
tillitis-{gateware,firmware,qemu}: init at 23.03.1

### PR DESCRIPTION
**Pull request by ghost**
- `2023-Apr-21`: the emulator [works](https://github.com/NixOS/nixpkgs/pull/227404#issuecomment-1518359567)! :tada: 

#### What is this?

[Tillitis](https://www.tillitis.se/) is an ~owner-controlled~ (Edit: apparently [I have been hoodwinked](https://github.com/NixOS/nixpkgs/pull/227404#issuecomment-1524106773)) cryptographic token, created by [Mullvad](https://mullvad.net/en/blog/2022/9/19/mullvad-creates-a-hardware-company/).

#### Description of changes

- [x] All of it builds using our `clang`, `qemu`, `sdcc`, and FPGA circuit synthesis tools (`yosys`/`nextpnr`) instead of the  dockerified toolchain that Tillitis provides.
- [x] The emulator [works](https://github.com/NixOS/nixpkgs/pull/227404#issuecomment-1518359567)
- [ ] Try the gateware on real hardware (mine cleared customs this week, arrived yesterday)

I'm probably abusing `pkgsCross` here.  But I can't get a full `riscv32-none-elf`-targeted `clang` to compile, even after spending a *lot* of time fighting with `llvmPackages.libstdcxx` (needed for a full stdenv bootstrap even though Tillitis doesn't use C++).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] powerpc64le-linux
  - [ ] aarch64-linux
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by amjoseph-nixpkgs** [comment]
The emulator works!

```
$ nix build -f . -L on_kgpe.tillitis-qemu.passthru.run && ./result
char device redirected to /dev/pts/19 (label chrid)
Hello, I'm firmware with tk1_name0:0x746b3120 tk1_name1:0x6d6b6466 tk1_version:0x00000001
```

File: pkgs/applications/virtualization/qemu/default.nix
Comment (Artturin) [inline pkgs/applications/virtualization/qemu/default.nix:1]: > This allows other packages to replace the `src` attribute, for
example in order to package a downstream fork of qemu.

this change isn't required for that

Comment (amjoseph-nixpkgs) [reply pkgs/applications/virtualization/qemu/default.nix:1]: Yeah not sure what i was thinking there.

File: pkgs/misc/tillitis/firmware-usb2serial.nix [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+  sourceRoot = "source/hw/boards/mta1-usb-v1/ch552_fw";
+
+  hardeningDisable = [ "all" ];
+
+  enableParallelBuilding = true;
+
+  buildInputs = [
+    sdcc
+  ];
```
Comment (Artturin) [inline pkgs/misc/tillitis/firmware-usb2serial.nix:22]: not nativeBuildInputs?

**Comment by mchack-work** [comment]
You won't be able to try the FPGA bitstream (what you call gateware) on an ordinary TKey bought from the Tillitis shop. It has already been provisioned with a bitstream inside the FPGA's own configuration memory and locked down.

Please ping me at mc at tillitis.se and I'll see what I can do.

**Comment by amjoseph-nixpkgs** [comment]
> and locked down.

I assume this means you've blown some kind of "no more programming" eFuse, rather than letting the user do so themselves.

Is this mentioned anywhere in the documentation?

I feel somewhat bait-and-switched here.  How is the user supposed to know what bitstream is on the FPGA if they can't load it themselves?  Since the FPGA programming pins aren't accessible via USB and these ICE FPGAs are not tamper-proof devices (i.e. they are not secure against physical attack) I'm not sure what threat model this is intended to protect against.

**Comment by RaitoBezarius** [comment]
I convinced 2 friends to buy 2 of them (and myself 2 of them) because of this feature… :C.

**Comment by JulienMalka** [comment]
> I convinced 2 friends to buy 2 of them (and myself 2 of them) because of this feature… :C.

That's a bit of a shame, me too...

**Comment by amjoseph-nixpkgs** [comment]
Wow, Tillitis [came through](https://github.com/tillitis/tillitis-key1/pull/119#issuecomment-1712426239); they're selling unlocked TKeys through their storefront.  I must be getting jaded; I had long ago stopped expecting that level of integrity from hardware vendors.  Very impressed.

It may be a while before I get back to pushing this PR forward, but I am hacking on this material locally again.

**Comment by bbigras** [comment]
is anyone working on packaging the SSH Agent for TKey? I'm stuck at trying to link monocypher with tkey-device-signer

**Pull request closed without merging**